### DISCUSSION TITLE
Share input/output runtime validations between server and client

### DISCUSSION THREAD

**Discussion by yukimotochern** [trpc/trpc#3048] [RFC / Ideas]
I agree with #1517 that we should have runtime check for the below reasons:

1. Some team members could potentially return objects of any type or make part of the returned object type any in the query or mutation. This kind of error will not be thrown in CI.
2. With wrong data structure being set to the client state, the whole client could break even for a minor api, which could have been handled gracefully if runtime validation was performed.

Edit because the discussion is locked:

I agree that any type is a bug of backend. But even in a mono-repo setup, the client/backend may still be maintained by different team. The client side runtime validation will limit the scope of a backend error by throwing it early. Normally, the client have to write the validation again. But the reuse of validation code could happen if there is another runtime safe layer in the mono repo setup much like the graphQL schema but in tRPC language. This 'contract' layer describes the input/output interface but with type 'and' runtime safeguard( the data transformer can sit in this layer too). I think that will be great!

**Comment by KATT** [comment]
We will not add runtime checks for the clients.

- If you want guaranteed safety of your returns on the backend, use procedure [output validation](https://trpc.io/docs/v10/output-validation) which will make these situations into internal server errors
- You should be able to trust what your server that likely sits in the same monorepo
- If you use `any` or send the wrong data structure, that's not tRPC's issue - especially not the _client_, that's a bug on your backend
